onnxruntime/contrib_ops/cuda/transformers/generation_device_helper.cc

```diff
@@ -721,40 +721,102 @@ Status PickGptPastState(const std::vector<OrtValue>& last_outputs,
-                                                              block_size_per_beam);
+    gsl::span<T> past_state_buffer = beam_state->GetPastStateBuffer();
 
-      gsl::span<T> past_key = past_span.subspan(j * block_size_per_beam, block_size_per_beam);
-      gsl::span<T> past_value = past_span.subspan(past_key_size + j * block_size_per_beam, block_size_per_beam);
-      CUDA_RETURN_IF_ERROR(cudaMemcpyAsync(past_key.data(), present_key.data(), present_key.size_bytes(),
-                                           cudaMemcpyDeviceToDevice, cuda_stream));
-      CUDA_RETURN_IF_ERROR(cudaMemcpyAsync(past_value.data(), present_value.data(), present_value.size_bytes(),
-                                           cudaMemcpyDeviceToDevice, cuda_stream));
+    for (int i = 0; i < num_present_tensors; ++i) {
+      gsl::span<T> current_layer_past_state_buffer = past_state_buffer.subspan(i * 2 * batch_beam_size *
+                                                                               num_heads * max_seq_length *
+                                                                               head_size);
+
+      // Create a tensor with same shape as the present state (use `representative_shape`).
+      OrtValue past;
+
+      // OrtValue doesn't own the buffer - so no need to pass in the allocator
+      Tensor::InitOrtValue(past_type, representative_shape, current_layer_past_state_buffer.data(), allocator->Info(), past);
+
+      next_inputs[gpt_subgraph_first_past_input_idx + i] = past;
     }
 
-    next_inputs[gpt_subgraph_first_past_input_idx + i] = past;
+    // Copy the selected beam indices to the device to launch the PickGptPastStateKernel kernel
+    CUDA_RETURN_IF_ERROR(cudaMemcpyAsync(beam_state->selected_next_indices.data(), beam_indices.data(),
+                                         beam_indices.size_bytes(), cudaMemcpyHostToDevice, cuda_stream));
+
+    typedef typename ToCudaType<T>::MappedType CudaT;
+    bool is_kernel_launch_successful = cuda::PickGptPastStateKernel<CudaT>(
+        reinterpret_cast<CudaT*>(past_state_buffer.data()),
+        // Use the buffer pointer of the first present state - the pre-allocated buffer is shared between
+        // the present states of all layers
+        reinterpret_cast<const CudaT*>(last_outputs[gpt_subgraph_first_present_output_idx].Get<Tensor>().Data<T>()),
+        beam_state->selected_next_indices.data(),
+        batch_beam_size,
+        past_seq_length,
+        max_seq_length,
+        num_heads,
+        head_size,
+        num_present_tensors,  // because num_present_tensors == num_layers
+        max_threads_per_block,
+        cuda_stream);
+
+    // TODO(hasesh): We should not run into a case where this is false for now.
+    // When we do plan to support a model that runs into this, we can enhance the
+    // PickGptPastStateKernel kernel.
+    ORT_ENFORCE(is_kernel_launch_successful, "Picking the past state for this GPT2 flavor is not supported yet");
```

Comment:
The cause is kernel not implemented for large hidden_size (> 2048). Maybe change the message to be more specified.